File: python/src/brain_api_quickstart/webhook_listener.py
"""Flask webhook listener with HMAC verification.

Binds to a port, exposes POST /webhook, verifies the X-Revmachina-Signature
header against the subscription's signing secret, then invokes the caller's
on_event handler.
"""

from __future__ import annotations

import threading
from typing import Callable

from flask import Flask, request
from werkzeug.serving import make_server

from .client import verify_webhook_signature


def start_webhook_listener(
    port: int,
    signing_secret: str,
    on_event: Callable[[dict, dict], None],
) -> "_ListenerHandle":
    """Start a background-thread Flask listener.

    on_event(event_payload: dict, headers: {"deliveryId": str, "eventType": str})
    """
    app = Flask(__name__)

    @app.post("/webhook")
    def receive():  # type: ignore[no-untyped-def]
        sig_header = request.headers.get("X-Revmachina-Signature", "")
        delivery_id = request.headers.get("X-Revmachina-Delivery-Id", "")
        event_type = request.headers.get("X-Revmachina-Event", "")
        payload = request.get_data(as_text=True)

        if not verify_webhook_signature(signing_secret, payload, sig_header):
            print(f"  [webhook] HMAC verification FAILED — rejecting delivery {delivery_id}")
            return ("invalid signature", 401)

        try:
            event = request.get_json(force=True, silent=False)
        except Exception:
            return ("invalid json", 400)

        on_event(event or {}, {"deliveryId": delivery_id, "eventType": event_type})
        return ("ok", 200)

    server = make_server("0.0.0.0", port, app)
    thread = threading.Thread(target=server.serve_forever, daemon=True)
    thread.start()
    return _ListenerHandle(server, thread)


class _ListenerHandle:
    def __init__(self, server, thread) -> None:  # type: ignore[no-untyped-def]
        self._server = server
        self._thread = thread

    def close(self) -> None:
        self._server.shutdown()
        self._thread.join(timeout=2.0)

File: typescript/src/webhook-listener.ts
// Express webhook listener with HMAC verification.
// Binds to a port, exposes POST /webhook, verifies the X-Revmachina-Signature
// header against the subscription's signing secret, then invokes the caller's
// onEvent handler.

import express from 'express'
import type { Request, Response } from 'express'
import { verifyWebhookSignature } from './client.js'

export interface WebhookEvent {
  event_id:    string
  event_type:  string
  tenant_id:   string
  lead_id:     string | null
  occurred_at: string
  data:        Record<string, unknown>
}

export function startWebhookListener(opts: {
  port:           number
  signingSecret:  string
  onEvent:        (event: WebhookEvent, headers: { deliveryId: string; eventType: string }) => void
}): { close: () => Promise<void> } {
  const app = express()
  // RAW body required so we can verify HMAC over the exact bytes the server signed.
  app.use(express.raw({ type: 'application/json' }))

  app.post('/webhook', (req: Request, res: Response) => {
    const sigHeader  = String(req.header('X-Revmachina-Signature') ?? '')
    const eventType  = String(req.header('X-Revmachina-Event') ?? '')
    const deliveryId = String(req.header('X-Revmachina-Delivery-Id') ?? '')
    const payload    = (req.body as Buffer).toString('utf-8')

    const valid = verifyWebhookSignature({
      signingSecret:   opts.signingSecret,
      payload,
      signatureHeader: sigHeader,
    })
    if (!valid) {
      console.error(`  [webhook] HMAC verification FAILED — rejecting delivery ${deliveryId}`)
      return res.status(401).send('invalid signature')
    }

    let event: WebhookEvent
    try {
      event = JSON.parse(payload) as WebhookEvent
    } catch {
      return res.status(400).send('invalid json')
    }

    opts.onEvent(event, { deliveryId, eventType })
    res.status(200).send('ok')
  })

  const server = app.listen(opts.port)
  return {
    close: () =>
      new Promise<void>((resolve) => {
        server.close(() => resolve())
      }),
  }
}
